subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java

```diff
@@ -78,25 +89,25 @@ ProviderInternal<JavaToolchain> findMatchingToolchain(JavaToolchainSpec filter)
         return new DefaultProvider<>(() -> resolveToolchain(filterInternal));
     }
 
-    @Nullable
-    private JavaToolchain resolveToolchain(JavaToolchainSpecInternal filterInternal) throws Exception {
-        filterInternal.finalizeProperties();
+    private JavaToolchain resolveToolchain(JavaToolchainSpecInternal requestedSpec) throws Exception {
+        requestedSpec.finalizeProperties();
 
-        if (!filterInternal.isValid()) {
+        if (!requestedSpec.isValid()) {
             throw DocumentedFailure.builder()
                 .withSummary("Using toolchain specifications without setting a language version is not supported.")
                 .withAdvice("Consider configuring the language version.")
                 .withUpgradeGuideSection(7, "invalid_toolchain_specification_deprecation")
                 .build();
         }
 
-        if (!filterInternal.isConfigured()) {
-            return null;
-        }
+        boolean useFallback = !requestedSpec.isConfigured();
+        JavaToolchainSpecInternal actualSpec = useFallback ? fallbackToolchainSpec : requestedSpec;
+        // We can't use the key of the fallback toolchain spec, because it is a spec that can match configured requests as well
+        JavaToolchainSpecInternal.Key actualKey = useFallback ? FALLBACK_TOOLCHAIN_KEY : requestedSpec.toKey();
 
-        Object resolutionResult = matchingToolchains.computeIfAbsent(filterInternal.toKey(), key -> {
+        Object resolutionResult = matchingToolchains.computeIfAbsent(actualKey, key -> {
```

Comment:
Why do we need to go through the map here? Can't we define a constant in this class for the toolchain from the current JVM and return that as toolchain if nothing is configured? Or would that complicate things?